splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java

```diff
@@ -29,108 +29,114 @@
-					to=_instance=new TimestampOracle(timestampBlockManager,blockSize);
-				}
-			}
-		}
-		return to;
-	}
-	
-	private TimestampOracle(TimestampBlockManager timestampBlockManager, int blockSize) throws TimestampIOException {
+        TimestampOracle to = _instance;
+        if(to==null){
+            synchronized(TimestampOracle.class){
+                to = _instance;
+                if(to==null){
+                    SpliceLogUtils.info(LOG,"Initializing TimestampOracle...");
+                    to=_instance=new TimestampOracle(timestampBlockManager,blockSize);
+                }
+            }
+        }
+        return to;
+    }
+
+    private TimestampOracle(TimestampBlockManager timestampBlockManager, int blockSize) throws TimestampIOException {
         this.timestampBlockManager=timestampBlockManager;
         this.blockSize = blockSize * TIMESTAMP_INCREMENT;
-		initialize();
-	}
+        initialize();
+    }
 
     /**
      * Read the current state of the block from the timestampBlockManager
      *
      * @throws TimestampIOException
      */
-	private void initialize() throws TimestampIOException {
-			synchronized(this) {
+    private void initialize() throws TimestampIOException {
+            synchronized(this) {
                 _maxReservedTimestamp = timestampBlockManager.initialize();
-				_timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
-			}
-			try {
-				registerJMX();
-			} catch (Exception e) {
+                _timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
+            }
+            try {
+                registerJMX();
+            } catch (Exception e) {
                 SpliceLogUtils.error(LOG, "Unable to register Timestamp Generator with JMX. Service will function but metrics will not be available.");
-			}
+            }
     }
 
-	public long getNextTimestamp(boolean refresh, boolean increment) throws TimestampIOException {
-		if (refresh) {
-            _maxReservedTimestamp = timestampBlockManager.initialize();
-			_timestampCounter.set(_maxReservedTimestamp);
-		}
-
-		long nextTS = increment ? _timestampCounter.addAndGet(TIMESTAMP_INCREMENT): _timestampCounter.get();
-		long maxTS = _maxReservedTimestamp; // avoid the double volatile read
-		if (nextTS > maxTS) {
-			reserveNextBlock(maxTS);
-		}
-		_numTimestampsCreated.incrementAndGet(); // JMX metric
-		return nextTS;
-	}
-
-    private void reserveNextBlock(long priorMaxReservedTimestamp) throws TimestampIOException {
-        synchronized(this)  {
-            if (_maxReservedTimestamp > priorMaxReservedTimestamp) return; // some other thread got there first
+    public long getCurrentTimestamp() {
+        return _timestampCounter.get();
+    }
+
+    public synchronized void bumpTimestamp(long timestamp) throws TimestampIOException {
+        if (timestamp > _timestampCounter.get()) {
+            timestampBlockManager.reserveNextBlock(timestamp);
+            _maxReservedTimestamp = timestamp;
```

Comment:
By calling **reserveNextBlock(timestamp)** we have guaranteed that **timestamp <= _maxReservedTimestamp**. Reassigning **_maxReservedTimestamp** to  **timestamp** doesn't look right.

Reply:
I'm not quite sure what else to put instead. @jyuanca, can you weigh in?